src/node_report.cc

```diff
@@ -0,0 +1,1029 @@
+#include <execinfo.h>
+#endif
+#include <sys/utsname.h>
+#endif
+
+#ifdef __APPLE__
+#include <mach-o/dyld.h>  // _dyld_get_image_name()
+#endif
+
+
+
+#ifndef _WIN32
+extern char** environ;
+#endif
+
+namespace nodereport {
+
+using node::arraysize;
+using node::environ_mutex;
+using node::Environment;
+using node::MallocedBuffer;
+using node::Mutex;
+using node::NativeSymbolDebuggingContext;
+using node::PerIsolateOptions;
+using v8::Function;
+using v8::HeapSpaceStatistics;
+using v8::HeapStatistics;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::Number;
+using v8::StackTrace;
+using v8::String;
+using v8::V8;
+using v8::Value;
+
+
+// Internal/static function declarations
+static void WriteNodeReport(Isolate* isolate,
+                            Environment* env,
+                            const char* message,
+                            const char* location,
+                            const std::string& filename,
+                            std::ostream& out,
+                            MaybeLocal<Value> error,
+                            TIME_TYPE* time);
+static void PrintVersionInformation(std::ostream& out, int* intend);
+static void PrintJavaScriptStack(std::ostream& out,
+                                      Isolate* isolate,
+                                      MaybeLocal<Value> error,
+                                      const char* location,
+                                      int* intend);
+static void PrintNativeStack(std::ostream& out, int* intend);
+#ifndef _WIN32
+static void PrintResourceUsage(std::ostream& out, int* intend);
+#endif
+static void PrintGCStatistics(std::ostream& out, Isolate* isolate, int* intend);
+static void PrintSystemInformation(std::ostream& out, int* intend);
+static void PrintLoadedLibraries(std::ostream& out, int* intend);
+static void PrintComponentVersions(std::ostream& out, int* intend);
+
+// Global variables
+static std::atomic_int seq = {0};  // sequence number for report filenames
+time_t boot_time;  // Node.js boot time absolute
+
+/*******************************************************************************
+ * External function to trigger a node report, writing to file.
+ *
+ * The 'name' parameter is in/out: an input filename is used if supplied, and
+ * the actual filename is returned.
+ ******************************************************************************/
+void TriggerNodeReport(Isolate* isolate,
+                       Environment* env,
+                       const char* message,
+                       const char* location,
+                       char* name,
+                       MaybeLocal<Value> error) {
+  std::ostringstream oss;
+  std::string filename;
+  std::shared_ptr<PerIsolateOptions> options;
+  if (env != nullptr) options = env->isolate_data()->options();
+
+  // Obtain the current time and the pid (platform dependent)
+  TIME_TYPE tm_struct;
+#ifdef _WIN32
+  DWORD pid;
```

Comment:
Can we use the same sort of typedef for pid as `TIME_TYPE` for time?
Something like
```c++
#ifdef _WIN32
typedef DWORD PID_TYPE;
#else  // UNIX, OSX
typedef pid_t PID_TYPE;
#endif
```
IIUC

Reply:
done.